### PR TITLE
Jormun: implement magic algo and tests

### PR DESCRIPTION
**Pull request by xlqian**
http://jira.canaltp.fr/browse/NAVP-309 

Implementing the nb_max_journey: Given N journeys, try to find t journeys which cover sections as many as possible.

For more details explication:  
http://confluence.canaltp.fr/pages/viewpage.action?pageId=8620368

### CODE REVIEW

File: source/jormungandr/jormungandr/scenarios/journey_filter.py [hunk 1]
```diff
@@ -136,6 +136,7 @@ def _filter_similar_journeys(journeys, request):
                           .format(other=j1.internal_id if worst == j2 else j2.internal_id))
 
 
+
```
Comment (TeXitoi) [inline source/jormungandr/jormungandr/scenarios/journey_filter.py]: ?

File: source/jormungandr/jormungandr/scenarios/new_default.py [hunk 2]
```diff
@@ -33,11 +33,16 @@
...
+    fill_uris, gen_all_combin, get_pseudo_duration
 from navitiacommon import type_pb2, response_pb2, request_pb2
 from jormungandr.scenarios.qualifier import min_from_criteria, arrival_crit, departure_crit, \
     duration_crit, transfers_crit, nonTC_crit, trip_carac, has_no_car, has_car, has_pt, \
     has_no_bike, has_bike, has_no_bss, has_bss, non_pt_journey, has_walk, and_filters
+import numpy as np
+
+SECTION_TYPES_TO_RETAIN = {response_pb2.PUBLIC_TRANSPORT, response_pb2.STREET_NETWORK}
+JOURNEY_TYPES_TO_RETAIN = {'best', 'comfort', 'non_pt_walk', 'non_pt_bike', 'non_pt_bss'}
+STREE_NETWORK_MODE_TO_RETAIN = {response_pb2.Car, response_pb2.Bike, response_pb2.Bss}
```
Comment (TeXitoi) [inline source/jormungandr/jormungandr/scenarios/new_default.py]: with a final T to street?

File: source/jormungandr/jormungandr/scenarios/new_default.py [hunk 3]
```diff
@@ -177,11 +299,77 @@ def culling_journeys(resp, request):
...
     no remove done in debug
     """
+    logger = logging.getLogger(__name__)
+
     if request['debug']:
         return
 
-    if request["max_nb_journeys"] and len(resp.journeys) > request["max_nb_journeys"]:
-        del resp.journeys[request["max_nb_journeys"]:]
+    if not request["max_nb_journeys"] or request["max_nb_journeys"] > len(resp.journeys):
```
Comment (TeXitoi) [inline source/jormungandr/jormungandr/scenarios/new_default.py]: `>=` no?

File: source/jormungandr/jormungandr/scenarios/new_default.py [hunk 4]
```diff
@@ -177,11 +299,77 @@ def culling_journeys(resp, request):
...
+    Journey_2 : Line 14 -> Line 6 -> Bus 165
+    Journey_3 : Line 14 -> Line 8 -> Bus 165
+
+    The candidate pool will be like [Journey_2, Journey_3]
+    The sections set will be like set([Line 14, Line 6, Line 8, Bus 165])
+    """
+    candidates_pool, sections_set = _build_candidate_pool_and_sections_set(resp)
+
+    nb_journeys_must_have = len(resp.journeys) - len(candidates_pool)
+    logger.debug("There are {0} journeys we mush keep".format(nb_journeys_must_have))
```
Comment (TeXitoi) [inline source/jormungandr/jormungandr/scenarios/new_default.py]: must

File: source/jormungandr/jormungandr/scenarios/new_default.py [hunk 5]
```diff
@@ -177,11 +299,77 @@ def culling_journeys(resp, request):
...
+
+    nb_journeys_to_find = request["max_nb_journeys"] - nb_journeys_must_have
```
Comment (TeXitoi) [inline source/jormungandr/jormungandr/scenarios/new_default.py:346]: `nb_journeys_to_find == len(candidates_pool)`, no?

Comment (xlqian) [inline source/jormungandr/jormungandr/scenarios/new_default.py:346]: Absolutely!!

File: source/jormungandr/jormungandr/scenarios/utils.py [hunk 6]
```diff
@@ -286,3 +288,45 @@ def fill_uris(resp):
...
+    c[t] = n
+    c[t+1] = 0
+    j = 0
+    while j < t:
+        yield c[0:-2]
+        j = 0
+        while (c[j] + 1) == c[j+1]:
+            c[j] = j
+            j += 1
+        c[j] += 1
```
Comment (TeXitoi) [inline source/jormungandr/jormungandr/scenarios/utils.py]: lack end of line at the end of the file

File: source/jormungandr/jormungandr/scenarios/new_default.py (hunk 5 above)
Comment (xlqian) [inline source/jormungandr/jormungandr/scenarios/new_default.py:346]: hmmm, In fact no...

Comment (xlqian) [inline source/jormungandr/jormungandr/scenarios/new_default.py:346]: let's say Kraken returns 15 journeys, with 1 best, 1 comfort, which means we have 2 journeys must keep and 13 journeys in candidates_pool.  On the other hand, the user wants 5 journeys in max, so have to find only 3 journeys(`nb_journeys_to_find`) among 13 journeys(`len(candidates_pool)`) in candidate_pool...

File: source/jormungandr/jormungandr/scenarios/journey_filter.py (hunk 1 above)
Comment (xlqian) [inline source/jormungandr/jormungandr/scenarios/journey_filter.py]: PEP8 convention says 
> Surround top-level function and class definitions with two blank lines. 

I do it as soon as I modify a function

File: source/jormungandr/jormungandr/scenarios/new_default.py (hunk 2 above)
Comment (xlqian) [inline source/jormungandr/jormungandr/scenarios/new_default.py]: ?

File: source/jormungandr/jormungandr/scenarios/new_default.py (hunk 3 above)
Comment (xlqian) [inline source/jormungandr/jormungandr/scenarios/new_default.py]: done

File: source/jormungandr/jormungandr/scenarios/new_default.py (hunk 4 above)
Comment (xlqian) [inline source/jormungandr/jormungandr/scenarios/new_default.py]: done

File: source/jormungandr/jormungandr/scenarios/utils.py (hunk 6 above)
Comment (xlqian) [inline source/jormungandr/jormungandr/scenarios/utils.py]: ok

**Comment by antoine-de** [comment]
Jenkins, test this please

File: source/jormungandr/jormungandr/scenarios/new_default.py (hunk 2 above)
Comment (xlqian) [inline source/jormungandr/jormungandr/scenarios/new_default.py]: Ok! I see

**Comment by xlqian** [comment]
Jenkins, test this please

File: source/jormungandr/jormungandr/scenarios/new_default.py [hunk 7]
```diff
@@ -170,18 +175,220 @@ def tag_journeys(resp):
...
+
+    The candidate pool will be like [Journey_2, Journey_3]
+    The sections set will be like set([Line 14, Line 6, Line 8, Bus 165])
+    """
+    candidates_pool, sections_set = _build_candidate_pool_and_sections_set(resp)
+
+    nb_journeys_must_have = len(resp.journeys) - len(candidates_pool)
+    logger.debug("There are {0} journeys we must keep".format(nb_journeys_must_have))
+    nb_journeys_to_find = request["max_nb_journeys"] - nb_journeys_must_have
+    if nb_journeys_to_find == 0:
```
Comment (TeXitoi) [inline source/jormungandr/jormungandr/scenarios/new_default.py]: What if nb_journeys_to_find is negative? For example, if max_nb_journeys == 1, we only want best

**Comment by TeXitoi** [comment]
OK for me.

**Pull request merged**